packetbeat/protos/sip/sip_message.go

```diff
@@ -0,0 +1,589 @@
+				separatedStrings := msg.separateCsv(lastElement)
+				for idx, element := range *separatedStrings {
+					if idx == 0 {
+						(*headers)[lastheader][lastHeaderEndIdx] = element
+					} else {
+						(*headers)[lastheader] = append((*headers)[lastheader], element)
+					}
+				}
+			}
+
+			// in case header line is NOT start [SP] or [HT]
+			// this line will be header parameter
+			// header parameter shuld be include ':'
+			// and split two string ,before first ':' and after first ':'.
+			headerKeyVal := strings.SplitN(string(msg.raw[s:e]), ":", 2)
+			key := strings.ToLower(strings.TrimSpace(headerKeyVal[0]))
+			val := strings.TrimSpace(headerKeyVal[1])
+
+			// in case string was not included the ':', it is not valid data, ignored.
+			if val == "" {
+				continue
+			}
+
+			// Convert compact form to full form
+			if len(key) == 1 {
+				switch key {
+				// referfenced by https://www.iana.org/assignments/sip-parameters/sip-parameters.xhtml
+				case "a":
+					key = "accept-contact" //[RFC3841]
+				case "b":
+					key = "referred-by" //[RFC3892]
+				case "c":
+					key = "content-type" //[RFC3261]
+				case "d":
+					key = "request-disposition" //[RFC3841]
+				case "e":
+					key = "content-encoding" //[RFC3261]
+				case "f":
+					key = "from" //[RFC3261]
+				case "i":
+					key = "call-id" //[RFC3261]
+				case "j":
+					key = "reject-contact" //[RFC3841]
+				case "k":
+					key = "supported" //[RFC3261]
+				case "l":
+					key = "content-length" //[RFC3261]
+				case "m":
+					key = "contact" //[RFC3261]
+				case "o":
+					key = "event" //[RFC6665] [RFC6446]
+				case "r":
+					key = "refer-to" //[RFC3515]
+				case "s":
+					key = "subject" //[RFC3261]
+				case "t":
+					key = "to" //[RFC3261]
+				case "u":
+					key = "allow-events" //[RFC6665]
+				case "v":
+					key = "via" //[RFC3261] [RFC7118]
+				case "x":
+					key = "session-expires" //[RFC4028]
+				case "y":
+					key = "identity" //[RFC8224]
+				}
+			}
+
+			// Initialize and add to map, if first find the header name in process
+			_, ok := (*headers)[key]
+			if !ok {
+				(*headers)[key] = []common.NetString{}
+			}
+
+			(*headers)[key] = append((*headers)[key], common.NetString(val))
+			lastheader = key
+		}
+		// in case last processed headers line is separated with comma.
+		// FIXME: this process is same to "Above process"
+		if lastheader != "" {
+			lastHeaderEndIdx := len((*headers)[lastheader]) - 1
+			if lastHeaderEndIdx < 0 {
+				continue
+			} // This case is not exist, maybe...
+
+			lastElement := string(getLast((*headers)[lastheader]))
+			separatedStrings := msg.separateCsv(lastElement)
+			for idx, element := range *separatedStrings {
+				if idx == 0 {
+					(*headers)[lastheader][lastHeaderEndIdx] = element
+				} else {
+					(*headers)[lastheader] = append((*headers)[lastheader], element)
+				}
+			}
+		}
+	}
+	return headers, firstLines
+}
+
+// TODO:The procedure with Content-Encoding(RFC3261).
```

Comment:
After merging this implementation let's create issues for the remaining TODOs and possibly document any limitations if they are likely to be encountered.